cloudinit/sources/helpers/azure.py

```diff
@@ -1055,4 +1055,212 @@ def dhcp_log_cb(out, err):
     )
 
 
+class BrokenAzureDataSource(Exception):
+    pass
+
+
+class NonAzureDataSource(Exception):
+    pass
+
+
+class OvfEnvXml:
+    NAMESPACES = {
+        "ovf": "http://schemas.dmtf.org/ovf/environment/1",
+        "wa": "http://schemas.microsoft.com/windowsazure",
+    }
+
+    def __init__(
+        self,
+        *,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        hostname: Optional[str] = None,
+        custom_data: Optional[bytes] = None,
+        disable_ssh_password_auth: Optional[bool] = None,
+        public_keys: Optional[List[dict]] = None,
+        preprovisioned_vm: bool = False,
+        preprovisioned_vm_type: Optional[str] = None,
+    ) -> None:
+        self.username = username
+        self.password = password
+        self.hostname = hostname
+        self.custom_data = custom_data
+        self.disable_ssh_password_auth = disable_ssh_password_auth
+        self.public_keys: List[dict] = public_keys or []
+        self.preprovisioned_vm = preprovisioned_vm
+        self.preprovisioned_vm_type = preprovisioned_vm_type
+
+    def __eq__(self, other) -> bool:
+        return self.__dict__ == other.__dict__
+
+    def __repr__(self) -> str:
+        props = [
+            "username=%s" % self.username,
+            "password=%s" % self.password,
+            "hostname=%s" % self.hostname,
+            "custom_data=%r" % self.custom_data,
+            "disable_ssh_password_auth=%s" % self.disable_ssh_password_auth,
+            "public_keys=%s" % self.public_keys,
+            "preprovisioned_vm=%s" % self.preprovisioned_vm,
+            "preprovisioned_vm_type=%s" % self.preprovisioned_vm_type,
+        ]
+        return "OvfEnvXml(" + " ".join(props) + ")"
+
+    @classmethod
+    def parse_text(cls, ovf_env_xml: str) -> "OvfEnvXml":
+        """Parser for ovf-env.xml data.
+
+        :raises NonAzureDataSource: if XML is not in Azure's format.
+        :raises BrokenAzureDataSource: if XML is unparseable or invalid.
+        """
+        try:
+            root = ElementTree.fromstring(ovf_env_xml)
+        except Exception as e:
+            error_str = "Invalid ovf-env.xml: %s" % e
+            raise BrokenAzureDataSource(error_str) from e
```

Comment:
We should report a diagnostic event for invalid ovf-env.xml

Reply:
It does get reported in crawl_metadata 👍